File: src/Web/JavaScriptJS/prototype.js
'use strict';

const serviceProfilePrototype = {
  clone() {
    return Object.assign(Object.create(Object.getPrototypeOf(this)), this, {
      features: [...this.features],
    });
  },

  describe() {
    return `${this.name}: ${this.features.join(',')}`;
  },
};

function createServiceProfile(name, features) {
  return Object.assign(Object.create(serviceProfilePrototype), {
    name,
    features: [...features],
  });
}

const original = createServiceProfile('orders', ['metrics']);
const canary = original.clone();

canary.name = 'orders-canary';
canary.features.push('tracing');

console.log(`original=${original.describe()}`);
console.log(`clone=${canary.describe()}`);

File: src/Scripting/RubyRB/prototype.rb
class ServiceProfile
  attr_accessor :name, :features

  def initialize(name, features)
    @name = name
    @features = features
  end

  def initialize_copy(source)
    super
    @features = source.features.dup
  end

  def describe
    "#{name}: #{features.join(',')}"
  end
end

original = ServiceProfile.new('orders', ['metrics'])
canary = original.dup
canary.name = 'orders-canary'
canary.features << 'tracing'

puts "original=#{original.describe}"
puts "clone=#{canary.describe}"
